Cloud cover benchmark tutorial

klwetstone opened the pull request:
- Add a folder in `competition` for `cloud-cover`
- Make the benchmark tutorial available, with some tweaks from the [blog post](https://www.drivendata.co/blog/cloud-cover-benchmark/) version to run directly in the PC hub

TomAugspurger commented on competitions/cloud-cover/benchmark-tutorial.ipynb:
```diff
@@ -0,0 +1,2586 @@
+   "outputs": [],
+   "source": [
+    "%%capture\n",
+    "!pip install pandas_path pytorch_lightning cloudpathlib loguru typer segmentation_models_pytorch"
```
I'm deploying an update today with pytorch-lighting, and segmentation-models-pytorch should already be available. So I think those two can be dropped.

klwetstone replied:
Sounds good, just let me know when that's ready (I'm testing it out now, and looks like the current version still needs them installed)

TomAugspurger commented on competitions/cloud-cover/benchmark-tutorial.ipynb:
```diff
@@ -0,0 +1,2586 @@
+    }
+   ],
+   "source": [
+    "example_chip = example_chip.iloc[0]\n",
```
Can you avoid reassigning this to the same variable? Previously this was a dataframe and now it's becoming a series, so users wouldn't be able to rerun this cell.

klwetstone replied:
great point!

TomAugspurger commented on competitions/cloud-cover/benchmark-tutorial.ipynb:
```diff
@@ -0,0 +1,2586 @@
+    },
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAAATsAAAE/CAYAAAA5TWTRAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8/fFQqAAAACXBIWXMAAAsTAAALEwEAmpwYAAEAAElEQVR4nOz9eayva5bXh33WM7zv+5v23uece+69NXR19QBtaAw2yEQRJmkZR3YGm8jCNk6I1Q6EKMQmshwZcByLWDgmkSMcObISx8YMHsBTYqPYJoGoTYgJICLHElMPdFd3Vdete8+0h9/vHZ5h5Y/1vL+97+1btw7dXdQtOE/p3Prt3/DOz3rW+q7v+i5RVd6MN+PNeDP+eh/uW30 ...
```
I'm seeing an `AttributeError: 'functools.partial' object has no attribute '__name__'` on this line when running this on the (staging) hub. Do you see it in prod?

IIRC this was from some incompatibility between some pytorch library and pandas. Do you recall how we solved it in the last notebook?

klwetstone replied:
I don't see this in prod (mine runs fine), but I have run into that error before and can't remember exactly how I fixed it. What are you versions of pytorch and pandas?

TomAugspurger commented on competitions/cloud-cover/benchmark-tutorial.ipynb:
```diff
@@ -0,0 +1,2586 @@
+    "    \"\"\"Given the path to the directory of Sentinel-2 chip feature images,\n",
+    "    plots the true color image\"\"\"\n",
+    "    chip_dir = data_dir / chip_id\n",
+    "    red = get_xarray(chip_dir / \"B04.tif\")\n",
```
get_xarray can be replaced with rioxarray.open_rasterio

```suggestion
    "    red = rioxarray.open_rasterio(chip_dir / \"B04.tif\").squeeze()\n",
```
likewise for the other two.

TomAugspurger reviewed (commented):
I have an update coming to the GPU environment shortly (likely later today) that'll have pytorch-lightning and torchgeo.

Do we have STAC items for the training data? I'd like to avoid all the helper functions like `lat_long_bounds`, `get_xarray`, `true_color_img`, etc. if possible. I'm worried about the long-term maintenance costs of having too much code in these notebooks. Any ideas you have to cut down on the amount of code here would be great to hear.

klwetstone commented:
Thanks for the feedback! Working on incorporating all the tweaks.

> Do we have STAC items for the training data?

Participants will be starting with the competition data we've provided (available in the attached `driven-data` volume) rather than STAC items, so we're mirroring that process to provide relevant helper code.

> I'd like to avoid all the helper functions like lat_long_bounds, get_xarray, true_color_img, etc. if possible. I'm worried about the long-term maintenance costs of having too much code in these notebooks. Any ideas you have to cut down on the amount of code here would be great to hear.

We currently have cells in the notebook to write out all of the source code to `benchmark_src` (`cloud_model.py`, `cloud_dataset.py`, `losses.py`, and `main.py`) starting with the "Build the model" section. We could save those in a separate scripts folder, remove the script code from the notebook, and just include a description and reference to the scripts. Would that cut it down enough?

I agree it's quite long :) since it's our only helper code document we like to be pretty comprehensive, but we can remove more from the planetary computer version if needed!

TomAugspurger commented on competitions/cloud-cover/benchmark-tutorial.ipynb:
```diff
@@ -0,0 +1,2516 @@
+    },
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAAATsAAAE/CAYAAAA5TWTRAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8/fFQqAAAACXBIWXMAAAsTAAALEwEAmpwYAAEAAElEQVR4nOz9eayva5bXh33WM7zv+5v23uece+69NXR19QBtaAw2yEQRJmkZR3YGm8jCNk6I1Q6EKMQmshwZcByLWDgmkSMcObISx8YMHsBTYqPYJoGoTYgJICLHElMPdFd3Vdete8+0h9/vHZ5h5Y/1vL+97+1btw7dXdQtOE/p3Prt3/DOz3rW+q7v+i5RVd6MN+PNeDP+eh/uW30 ...
```
I think we want

```
        self.save_hyperparameters(ignore=["x_train", "y_train", "x_val", "y_val"])
```

unless we want to include the training / validation data in the hyperparameters.

klwetstone replied:
I don't think they're included as is. It only stores the arguments under `self.hparams`, which in this case doesn't include the input data - `x_train`, `y_train`, `x_val`, and  `y_val` are all input directly to `CloudModel` rather than being included in the `hparams` dict.

TomAugspurger commented on competitions/cloud-cover/benchmark-tutorial.ipynb:
```diff
@@ -0,0 +1,2516 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "First, we'll download all of the competition data to a local folder. For instructions, see the competition's data download [page](https://www.drivendata.org/competitions/83/cloud-cover/data/). You'll need to join the competition in order to view the data page. \n",
```
Does this need to be updated to mention that the training data are available on the Planetary Computer at `/driven-data/cloud-cover`.

TomAugspurger commented:
> Participants will be starting with the competition data we've provided (available in the attached driven-data volume) rather than STAC items, so we're mirroring that process to provide relevant helper code.

Gotcha. I think that for future competitions, STAC items might be helpful. They're pretty handy to have around :)

Just one question for you about how to explain accessing the data at `/driven-data` on the Planetary Computer: https://github.com/microsoft/PlanetaryComputerExamples/pull/112#discussion_r766740210

I think the error I saw on staging was from a change in pandas. https://github.com/microsoft/PlanetaryComputerExamples/pull/112#pullrequestreview-828981572 has a suggestion for future-proofing the code, which I think is a sensible change anyways but I'm not too familiar with pytorch-lightning.

There's also a couple linting failures: https://github.com/microsoft/PlanetaryComputerExamples/runs/4476206460?check_suite_focus=true

1. Might need a `import pandas_path  # noqa`
2. Will need to change the level-1 headers to level-2, and the level-2 headers to level-3, so that this fits the header structure with our other notebooks / site.

rbavery commented:
@klwetstone Hi! Thanks for putting together this notebook. I'm participating in this challenge, would it be possible to include a requirements.txt or environment.yml with this tutorial? 

Setting up environments is tough and I don't necessarily expect to be able to install the environment from a one-liner when geospatial and deep learning dependencies are mixed, but it'd be helpful to see pinned versions you successfully used so that I can reproduce your notebook. I'm currently stuck on a pytorch lightning error that looks related to the pytoruch lightning version.

TomAugspurger commented:
At least for the Planetary Computer Hub, we have docker images and conda-lock files at https://github.com/microsoft/planetary-computer-containers.

klwetstone commented:
@rbavery good question. Our benchmark environments are designed to work in the planetary computer hub, so the resources on the hub that Tom provided should give you the right information about how the benchmark was run. We also have two environment `yml` files in the runtime [repository](https://github.com/drivendataorg/cloud-cover-runtime/tree/main/runtime) for the competition. For more details about the runtime, see the competition page on [Code submission format](https://www.drivendata.org/competitions/83/cloud-cover/page/412/).

For other competition questions, the [forum](https://community.drivendata.org/c/cloud-cover/50) is another good place to start. Hope that helps!

klwetstone commented:
@TomAugspurger ready for another round! 
- Added text explaining data storage in the Hub
- Demoted all headers by 1 (level-1 to level-2, etc.)
- Added `# noqa` next to the `pandas_path` import

I responded to your [comment](https://github.com/microsoft/PlanetaryComputerExamples/pull/112#discussion_r766739354) about the `self.save_hyperparameters` line. I’m not sure it needs to be changed, but happy to look into it more!

TomAugspurger commented on competitions/cloud-cover/benchmark-tutorial.ipynb:
```diff
@@ -0,0 +1,2518 @@
+    "import matplotlib.pyplot as plt\n",
+    "import numpy as np\n",
+    "import pandas as pd\n",
+    "from pandas_path import path # noqa\n",
```
```suggestion
    "from pandas_path import path  # noqa\n",
```

TomAugspurger replied:
Hmm I can't merge this suggestion from the GitHub UI. Strange. I'll make it locally and push it.

TomAugspurger replied:
Huh, that's not working either. So I'll need to you to make the change @klwetstone.

btw https://github.com/microsoft/PlanetaryComputerExamples/blob/main/CONTRIBUTING.md#linting has some short instructions if you want to run the linting checks locally.

klwetstone commented:
@TomAugspurger added that tweak, and ran the linting locally per your link. Should be good to go!

TomAugspurger commented:
Great, thanks!
